### ISSUE TITLE
Cannot put percolator query on geo_shape field

### ISSUE THREAD

**Issue by amarandon** [elastic/elasticsearch#10500]
Here is a script to reproduce the problem:

```shell
ES_SERVER=http://localhost:9200
curl -XPUT $ES_SERVER/myindex?pretty

# Insert mapping with geo_shape type
curl -XPUT $ES_SERVER/myindex/_mapping/mytype?pretty -d '
{
    "mytype" : {
        "properties": {
            "location": {
                "type": "geo_shape",
                "tree": "quadtree",
                "precision": "1m"
            }
        }
    }
}
'

# Check that mapping is correct
curl -XGET $ES_SERVER/myindex/_mapping/?pretty

curl -XPUT $ES_SERVER/myindex/mytype/mydocument?pretty -d'
{
    "location" : {
        "type" : "point",
        "coordinates" : [1.44207, 43.59959]
    }
}
'

# Check that we can retrieve our document with a normal query
curl -XGET $ES_SERVER/myindex/mytype/_search?pretty -d '
{
    "query": {
        "geo_shape": {
            "location": {
                "shape": {
                    "type": "envelope",
                    "coordinates": [[0, 50],[2, 40]]
                }
            }
        }
    }
}
'
# Try to submit the same query to the percolator. FAIL!
curl -XPUT $ES_SERVER/myindex/.percolator/myquery?pretty -d '
{
    "query": {
        "geo_shape": {
            "location": {
                "shape": {
                    "type": "envelope",
                    "coordinates": [[0, 50],[2, 40]]
                }
            }
        }
    }
}
'

```

Everything goes as expected except the last request which returns:
```json
{
  "error" : "PercolatorException[[myindex] failed to parse query [myquery]]; nested: QueryParsingException[[myindex] Field [location] is not a geo_shape]; ",
  "status" : 500
}
```

Following the documentation, it seems that it should have worked. I'll try to figure out the problem and submit a patch (whether it turns out to be a documentation patch or a code patch), but it would be great if someone could confirm if it's a bug or a misuse from my end.

**Comment by clintongormley** [comment]
Hi @amarandon 

I've just tried this on 1.5.0, and it works correctly. 

    DELETE myindex
    
    PUT /myindex?pretty
    
    # Insert mapping with geo_shape type
    PUT /myindex/_mapping/mytype?pretty
    {
      "mytype": {
        "properties": {
          "location": {
            "type": "geo_shape",
            "tree": "quadtree",
            "precision": "1m"
          }
        }
      }
    }
    
    # Check that mapping is correct
    GET /myindex/_mapping/?pretty
    
    
    PUT /myindex/mytype/mydocument?pretty
    {
      "location": {
        "type": "point",
        "coordinates": [
          1.44207,
          43.59959
        ]
      }
    }
    
    # Check that we can retrieve our document with a normal query
    
    GET /myindex/mytype/_search?pretty
    {
        "query": {
            "geo_shape": {
                "location": {
                    "shape": {
                        "type": "envelope",
                        "coordinates": [[0, 50],[2, 40]]
                    }
                }
            }
        }
    }
    
    # Try to submit the same query to the percolator. Works
    PUT /myindex/.percolator/myquery?pretty
    {
        "query": {
            "geo_shape": {
                "location": {
                    "shape": {
                        "type": "envelope",
                        "coordinates": [[0, 50],[2, 40]]
                    }
                }
            }
        }
    }
    
    # Percolate request works too
    POST myindex/mytype/_percolate
    {
      "doc": {
        "location": {
          "type": "point",
          "coordinates": [
            1.44207,
            43.59959
          ]
        }
      }
    }

**Comment by amarandon** [comment]
Hi @clintongormley Thanks for trying it out. I found out that the issue is triggered by having `index.percolator.map_unmapped_fields_as_string: true` in my config file. I'm in the process of migrating an app built against an earlier version and found out that I had to enable that option to keep it working because not all the percolator queries we record have corresponding mappings.

**Comment by amarandon** [comment]
@clintongormley Does it still for you with `index.percolator.map_unmapped_fields_as_string: true` in your config file?

**Comment by clintongormley** [comment]
@amarandon if you're using that option, then I'm not surprised it fails... A geoshape query can't work on a string field.  You need to have the field specified in the mapping before you can create a percolator which uses it.  Otherwise (with that setting enabled) it will assume that the missing field is a string, and... fail

**Comment by amarandon** [comment]
@clintongormley But I do have the field specified in the mapping before I create the percolator which uses it. In the test script I provided, we create that mapping explicitly and even check that it's been properly created with a GET request before trying to create a percolator query against it. In other words the geo_shape field is not unmapped and shouldn't be affected by that option.

**Comment by clintongormley** [comment]
Gotcha!  And I can recreate, too.  Agreed, this is a bug.

@martijnvg please could you take a look

Here's the full recreation:

    PUT /myindex?pretty
    {
      "settings": {
        "index.percolator.map_unmapped_fields_as_string":true
      }
    }
    
    # Insert mapping with geo_shape type
    PUT /myindex/_mapping/mytype?pretty
    {
      "mytype": {
        "properties": {
          "location": {
            "type": "geo_shape",
            "tree": "quadtree",
            "precision": "1m"
          }
        }
      }
    }
    
    # Check that mapping is correct
    GET /myindex/_mapping/?pretty
    
    
    PUT /myindex/mytype/mydocument?pretty
    {
      "location": {
        "type": "point",
        "coordinates": [
          1.44207,
          43.59959
        ]
      }
    }
    
    # Check that we can retrieve our document with a normal query
    
    GET /myindex/mytype/_search?pretty
    {
        "query": {
            "geo_shape": {
                "location": {
                    "shape": {
                        "type": "envelope",
                        "coordinates": [[0, 50],[2, 40]]
                    }
                }
            }
        }
    }
    
    # Try to submit the same query to the percolator. Works
    PUT /myindex/.percolator/myquery?pretty
    {
        "query": {
            "geo_shape": {
                "location": {
                    "shape": {
                        "type": "envelope",
                        "coordinates": [[0, 50],[2, 40]]
                    }
                }
            }
        }
    }
    
    # Percolate request works too
    POST myindex/mytype/_percolate
    {
      "doc": {
        "location": {
          "type": "point",
          "coordinates": [
            1.44207,
            43.59959
          ]
        }
      }
    }

**Comment by clintongormley** [comment]
This still fails in 2.2 and master, but it fails when trying to PUT the percolator query with `location is not a geoshape`.

@martijnvg are percolator queries still parse-once, or are they parsed on each execution? If the latter, then could we just remove the `index.percolator.map_unmapped_fields_as_string` setting?

**Comment by martijnvg** [comment]
@clintongormley Percolator queries are still parsed once. 

This issue was caused by a bug that if the 'map unmapped fields as strings' was enabled it would even substitute found fields with string fields! The `geo_shape` query has a hard check if what type of field is being returned from the mapping as therefor fails. Luckily this is easy to fix: #16043
